### ISSUE TITLE
Certain conditional types allow unsound assignments

### ISSUE THREAD

**Issue by mkantor** [microsoft/TypeScript#55733]
### 🔎 Search Terms

"conditional type" generic assignment

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about conditional types

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAQg9gDwDwBUB8UC8UDeUBuAhgDYCuEAXFClAL4BQ9okUAqgHYBGiAku+xABOqDNhoQEwCOwAmAZ1iIkAS3YAzIVABKGAPxsuvfkKQ6oVPKUMIIMqjQbNoHbsnRYDrvgOHoA2gDkVq62AQC6jDIQAMbEhILQalbRwMpw7FDBSugAFAj2AJRULtlojEnsKWkZEoQAtmDEECJ5hbj0UFBNwFCEHlkIeQUdvR4BwAAWygpyE3CkxDJQnNCENYKCcIIB9LRAA

### 💻 Code

```ts
type Box<T> = { value: T }

type UnboxInner<T> = T extends Box<infer R> ? UnboxInner<R> : { unboxed: T }
type Unbox<T> = UnboxInner<T>['unboxed']

declare function unbox<T>(x: T): Unbox<T>

function example<T>(x: T) {
  let a = unbox(x)
  a = 'this should be an error'
}
```

### 🙁 Actual behavior

In `example` we can assign arbitrary values to `a`, even though it's typed as `Unbox<T>`.

### 🙂 Expected behavior

Attempting to assign invalid values to `a` (just about anything, since its type depends on a type parameter) should cause a type error.

### Additional information about the issue

These issues may possibly be related:
- #52021
- #35533

cc @webstrand

**Comment by Andarist** [comment]
This is a fun one.

The constraint of the object type of this type:
```ts
type Unbox<T> = UnboxInner<T>['unboxed']
```
is computed as
```ts
type BaseConstraintOfObjectType = { unboxed: unknown; } | { unboxed: T; }
```

This makes sense because the constraint is computed based on both branches of this conditional type and the `infer R` is also "replaced" with its own constraint (and that is `unknown`).

So... later we unpack the `'unboxed'` property from this union, so we end up with `unknown | T` and that is reduced to just `unknown` (it's not that the reduction thing matters here though). And string can be related to `unknown`.

**Comment by GheorgheP** [comment]
But why is allowed the assignment to a variable of type `unknown`?
Isn't this the real issue?

**Comment by Andarist** [comment]
You can always assign things to their supertypes and `unknown` is like an ultimate supertype. Everything is assignable to `unknown` (except `never`). If you think about types in terms of sets then `unknown` is like a set that contains *everything*.

**Comment by GheorgheP** [comment]
Hmmm, how then the `any` type is described and at what point `any` is different to `unknown`?
Sure I don't know how to describe the `unknown` in terms of sets, but this form looks more like `any`.
Shouldn't `unknown` be something like bottom type?

**Comment by fatcerberus** [comment]
`any` means "this value won't be typechecked".  `unknown` means "any possible value" but doesn't disable typechecking (it's not assignable to anything except itself).

> Shouldn't `unknown` be something like bottom type?

No, the bottom type is `never`.

**Comment by aweebit** [comment]
The reason for this issue is the same as in #62665. The only solution I know of is to use type inference in the definition of `Unbox`:
```ts
type Unbox<T> = UnboxInner<T> extends { unboxed: infer U } ? U : never;
```
